File: 卡码网/Go/test23.go
// 23. 二叉树的高度
package main

import (
	"fmt"
	"strings"
)

type Node struct{
	data string
	left,right *Node
}

func Max(a,b int) int{
	if a>b {
		return a
	}
	return b
}

func Index(nodes []string,target string) int{
	res:=-1
	for index,value:=range nodes {
		if value==target{
			res=index
			break
		}
	}
	return res
}

//创建二叉树
func createTree(pre,in []string,preLeft,preRight,inLeft,inRight int) *Node{
	if preLeft>preRight||inLeft>inRight {
		return nil
	}
	if preLeft==preRight {
		return &Node{pre[preLeft],nil,nil}
	}
	rootIndex:=Index(in,pre[preLeft])
	root:=&Node{pre[preLeft],createTree(pre,in,preLeft+1,preLeft+rootIndex-inLeft,inLeft,rootIndex-1),createTree(pre,in,preLeft+rootIndex-inLeft+1,preRight,rootIndex+1,inRight)}
	return root
}

//获取树高
func getHeight(root *Node) int{
	if root==nil {
		return 0
	}
	return Max(getHeight(root.left),getHeight(root.right))+1
}

func main(){
	for {
		var n int
		_,err:=fmt.Scanf("%d\n",&n)
		if err!=nil {
			return
		}
		var preStr,inStr string
		fmt.Scanln(&preStr)
		fmt.Scanln(&inStr)
		preNodes:=strings.Split(preStr,"")
		inNodes:=strings.Split(inStr,"")
		root:=createTree(preNodes,inNodes,0,n-1,0,n-1)
		fmt.Println(getHeight(root))
	}
}

File: 卡码/KamaCoder/C++/test23.cpp
//23. 二叉树的高度
#include <iostream>
using namespace std;

struct Node{
	char data;
	Node* left;
	Node* right;
	Node(char data):data(data),left(nullptr),right(nullptr){}
};

int getHeight(Node* root){
	int leftHeight=0,rightHeight=0;
	if(root->left!=nullptr) leftHeight=getHeight(root->left);
	if(root->right!=nullptr) rightHeight=getHeight(root->right);
	return max(leftHeight,rightHeight)+1;
}

Node* createTree(string preOrder,string inOrder,int preLeft,int preRight,int inLeft,int inRight){
	if(preLeft>preRight||inLeft>inRight) return nullptr;
	int rootIndex=inOrder.find(preOrder[preLeft]);
	Node* root=new Node(preOrder[preLeft]);
	int leftNum=rootIndex-inLeft;
	root->left=createTree(preOrder,inOrder,preLeft+1,preLeft+leftNum,inLeft,rootIndex-1);
	root->right=createTree(preOrder,inOrder,preLeft+1+leftNum,preRight,rootIndex+1,inRight);
	return root;
}

int main(){
	int n;
	while(scanf("%d\n",&n)!=EOF){
		string preOrder,inOrder;
		cin>>preOrder>>inOrder;
		Node* root=createTree(preOrder,inOrder,0,n-1,0,n-1);
		printf("%d\n",getHeight(root));
	}
	return 0;
}